cluster/assemblestate/devices.go

```diff
@@ -0,0 +1,256 @@
+// handled, requeue them with [DeviceQueryTracker.RetryResponses].
+func (d *DeviceQueryTracker) ResponsesAvailable() <-chan struct{} {
+	return d.responses
+}
+
+// RetryResponses signals that device query responses should be retried. This
+// should be called when a response transmission fails.
+func (d *DeviceQueryTracker) RetryResponses() {
+	select {
+	case d.responses <- struct{}{}:
+	default:
+	}
+}
+
+// QueriesAvailable returns a channel that signals when there are device queries
+// ready to be sent to peers. After reading from this channel, it is expected
+// that all queries are handled via calls to [DeviceQueryTracker.QueryableFrom].
+// In the case that queries cannot be handled, requeue them with
+// [DeviceQueryTracker.RetryQueries].
+func (d *DeviceQueryTracker) QueriesAvailable() <-chan struct{} {
+	return d.queries
+}
+
+// RetryQueries signals that device queries should be retried. This should be
+// called when a query transmission fails.
+func (d *DeviceQueryTracker) RetryQueries() {
+	select {
+	case d.queries <- struct{}{}:
+	default:
+	}
+}
+
+// Query records a query from a peer for unknown device identities. If we have
+// identity information for the requested devices, a response will be queued.
+func (d *DeviceQueryTracker) Query(from DeviceToken, unknowns []DeviceToken) {
+	if d.unknowns[from] == nil {
+		d.unknowns[from] = make(map[DeviceToken]struct{})
+	}
+
+	changes := false
+	for _, u := range unknowns {
+		// if we don't know the requested device's id, then we drop the query
+		if _, ok := d.ids[u]; !ok {
+			continue
+		}
+
+		changes = true
+		d.unknowns[from][u] = struct{}{}
+	}
+
+	// if we got some queries we can answer, let the other side know there is
+	// new data to handle
+	if changes {
+		select {
+		case d.responses <- struct{}{}:
+		default:
+		}
+	}
+}
+
+// QueryResponses returns the device identities that should be sent to the
+// specified peer, along with an acknowledgment function that must be called
+// after successful transmission.
+func (d *DeviceQueryTracker) QueryResponses(from DeviceToken) (ids []Identity, ack func()) {
+	ids = make([]Identity, 0, len(d.unknowns[from]))
+	for rdt := range d.unknowns[from] {
+		id, ok := d.ids[rdt]
+		if !ok {
+			continue
+		}
+		ids = append(ids, id)
+	}
+
+	ack = func() {
+		for _, id := range ids {
+			delete(d.unknowns[from], id.RDT)
+		}
+	}
+
+	return ids, ack
+}
+
+// QueryableFrom returns unknown device RDTs that should be queried from the
+// specified source peer, along with an acknowledgment function that must be
+// called after successful transmission to track inflight queries.
+func (d *DeviceQueryTracker) QueryableFrom(source DeviceToken) (unknown []DeviceToken, ack func()) {
+	for rdt := range d.sources[source] {
+		if _, ok := d.ids[rdt]; ok {
+			continue
+		}
+
+		// if we have a query in-flight for this rdt, omit it from the queries
+		// we will send
+		if time.Since(d.inflight[rdt]) < d.timeout {
```

Comment:
Looking at this again `time.Since` would prevent us from mocking time but this is equivalent:
```suggestion
		if d.inflight[rdt].Add(d.timeout).After(timeNow()) {
```